lib/src/component_declaration/function_component.dart

```diff
@@ -0,0 +1,181 @@
+// FIXME right now only top level factory declarations will generate props configs.
+UiFactory<TProps> uiFunction<TProps extends UiProps>(
+  UiFunctionComponent<TProps> functionComponent,
+  FunctionComponentConfig<TProps> config, {
+  PropsFactory<TProps> propsFactory,
+  String displayName,
+}) {
+  if (config != null) {
+    if (propsFactory != null) {
+      throw ArgumentError('propsFactory cannot be used along with config');
+    }
+    propsFactory = config.propsFactory;
+    displayName ??= config.componentName;
+  }
+
+  // Get the display name from the inner function if possible so it doesn't become `_uiFunctionWrapper`
+  displayName ??= _getFunctionName(functionComponent);
+
+  dynamic _uiFunctionWrapper(Map props) {
+    return functionComponent(propsFactory.jsMap(props as JsBackedMap));
+  }
+
+  ReactDartFunctionComponentFactoryProxy factory =
+      react.registerFunctionComponent(
+    _uiFunctionWrapper,
+    displayName: displayName,
+  );
+
+  if (propsFactory == null) {
+    if (TProps != UiProps && TProps != _GenericUiProps) {
+      throw ArgumentError(
+          'config.propsFactory must be provided when using custom props classes');
+    }
+    propsFactory = PropsFactory.fromUiFactory(
+            ([backingMap]) => _GenericUiProps(factory, backingMap))
+        as PropsFactory<TProps>;
+  }
+
+  TProps _uiFactory([Map backingMap]) {
+    TProps builder;
+    if (backingMap == null) {
+      builder = propsFactory.jsMap(JsBackedMap());
+    } else if (backingMap is JsBackedMap) {
+      builder = propsFactory.jsMap(backingMap);
+    } else {
+      builder = propsFactory.map(backingMap);
+    }
+
+    return builder..componentFactory = factory;
+  }
+
+  return _uiFactory;
+}
+
+String _getFunctionName(Function function) {
+  return getProperty(function, 'name') ?? getProperty(function, '\$static_name');
+}
+
+class _GenericUiProps extends UiProps {
+  @override
+  final Map props;
+
+  _GenericUiProps(ReactComponentFactoryProxy componentFactory, [Map props])
+      : this.props = props ?? JsBackedMap() {
+    this.componentFactory = componentFactory;
+  }
+
+  @override
+  String get propKeyNamespace => '';
+
+  @override
+  bool get $isClassGenerated => true;
+}
+
+/// Helper class used to keep track of generated information for [uiFunction].
+@protected
+class FunctionComponentConfig<TProps extends UiProps> {
+  @protected
+  final PropsFactory<TProps> propsFactory;
+  final String componentName;
+
+  @protected
+  FunctionComponentConfig({this.propsFactory, this.componentName});
+}
+
+/// Helper class to keep track of props factories used by [uiFunction],
+class PropsFactory<TProps extends UiProps> {
+  /// Returns a typed props object backed by the specified [Map].
+  final TProps Function(Map props) map;
+
+  /// Returns a typed props object backed by the specified [JsBackedMap].
+  final TProps Function(JsBackedMap props) jsMap;
+
+  PropsFactory({
```

Comment:
Should this constructor be public? Is there a use case for an instance of this that is not created using the `fromUiFactory` factory constructor?

Reply:
This has applications for making over_react JS components wrappers